pkg/manager/member/pvc_cleaner.go

```diff
@@ -40,37 +49,161 @@ type PVCCleanerInterface interface {
 	skipReason := map[string]string{}
 
-	selector, err := label.New().Instance(tc.GetLabels()[label.InstanceLabelKey]).Selector()
+	pvcs, err := rpc.listAllPVCs(tc)
 	if err != nil {
-		return skipReason, fmt.Errorf("cluster %s/%s assemble label selector failed, err: %v", ns, tcName, err)
+		return skipReason, err
 	}
 
-	pvcs, err := rpc.pvcLister.PersistentVolumeClaims(ns).List(selector)
+	for _, pvc := range pvcs {
+		pvcName := pvc.GetName()
+		l := label.Label(pvc.Labels)
+		if !(l.IsPD() || l.IsTiKV()) {
+			skipReason[pvcName] = skipReasonPVCCleanerIsNotPDOrTiKV
+			continue
+		}
+
+		if pvc.Status.Phase != corev1.ClaimBound {
+			// If pvc is not bound yet, it will not be processed
+			skipReason[pvcName] = skipReasonPVCCleanerPVCNotBound
+			continue
+		}
+
+		if pvc.DeletionTimestamp != nil {
+			// PVC has been deleted, skip it
+			skipReason[pvcName] = skipReasonPVCCleanerPVCHasBeenDeleted
+			continue
+		}
+
+		if len(pvc.Annotations[label.AnnPVCDeferDeleting]) == 0 {
+			// This pvc has not been marked as defer delete PVC, can't reclaim the PV bound to this PVC
+			skipReason[pvcName] = skipReasonPVCCleanerIsNotDeferDeletePVC
+			continue
+		}
+
+		// PVC has been marked as defer delete PVC, try to reclaim the PV bound to this PVC
+		podName, exist := pvc.Annotations[label.AnnPodNameKey]
+		if !exist {
+			// PVC has not pod name annotation, this is an unexpected PVC, skip it
+			skipReason[pvcName] = skipReasonPVCCleanerPVCNotHasPodNameAnn
+			continue
+		}
+
+		_, err := rpc.podLister.Pods(ns).Get(podName)
+		if err == nil {
+			// PVC is still referenced by this pod, can't reclaim PV
+			skipReason[pvcName] = skipReasonPVCCleanerPVCeferencedByPod
+			continue
+		}
+		if !errors.IsNotFound(err) {
+			return skipReason, fmt.Errorf("cluster %s/%s get pvc %s pod %s from local cache failed, err: %v", ns, tcName, pvcName, podName, err)
+		}
+
+		// if pod not found in cache, re-check from apiserver directly to make sure the pod really not exist
+		_, err = rpc.kubeCli.CoreV1().Pods(ns).Get(podName, metav1.GetOptions{})
+		if err == nil {
+			// PVC is still referenced by this pod, can't reclaim PV
+			skipReason[pvcName] = skipReasonPVCCleanerPVCeferencedByPod
+			continue
+		}
+		if !errors.IsNotFound(err) {
+			return skipReason, fmt.Errorf("cluster %s/%s get pvc %s pod %s from apiserver failed, err: %v", ns, tcName, pvcName, podName, err)
+		}
+
+		// Without pd or tikv pod reference this defer delete PVC, start to reclaim PV
+		pvName := pvc.Spec.VolumeName
+		pv, err := rpc.pvLister.Get(pvName)
+		if err != nil {
+			if errors.IsNotFound(err) {
+				skipReason[pvcName] = skipReasonPVCCleanerNotFoundPV
+				continue
+			}
+			return skipReason, fmt.Errorf("cluster %s/%s get pvc %s pv %s failed, err: %v", ns, tcName, pvcName, pvName, err)
+		}
+
+		if pv.Spec.PersistentVolumeReclaimPolicy != corev1.PersistentVolumeReclaimDelete {
+			err := rpc.pvControl.PatchPVReclaimPolicy(tc, pv, corev1.PersistentVolumeReclaimDelete)
```

Comment:
if `pvReclaimPolicy` is set retain or user's storage class use Retain policy, should we delete the PV?

Reply:
I think it is necessary, no matter what reclaim policy PV starts with, we should reclaim it when it can be reclaimed, and this reclaim logic does not destroy the data security of normal PV. Just for those PVs that can be reclaimed.

Reply:
add a note in the comments of this option?

Reply:
Done